File: time-tracker/src/app/Ett.Common.Services/Services/Statistics/StatisticsService.cs
namespace Ett.Common.Services.Services.Statistics
{
    using System.Threading.Tasks;

    using Ett.Common.Dal.Infrastructure;
    using Ett.Common.Services.Models.Statistics;
    using Ett.Common.Services.Services.Common;
    using Ett.Common.Services.Services.Statistics.Interfaces;

    public abstract class StatisticsService<TRequest, TData>
        : UnitOfWorkService, IStatisticsService<TRequest, TData>
    {
        protected StatisticsService(IUnitOfWorkFactory unitOfWorkFactory)
            : base(unitOfWorkFactory)
        {
        }

        protected abstract Task<Statistics<TRequest, TData>> DoGetStatistics(IUnitOfWork uow, TRequest request);

        public virtual async Task<Statistics<TRequest, TData>> GetStatistics(TRequest request)
        {
            using (var uow = this.CreateWithDisabledLazyLoading())
            {
                return await this.DoGetStatistics(uow, request);
            }
        }
    }
}

File: query-builder/src/frontend/src/app/common/services/statistics/concrete/statistics.service.ts
import { HttpTask } from './../../../tasks/http-task';
import { Task } from './../../../tasks/task';
import { Injectable } from '@angular/core';
import { IStatisticsService } from '../i-statistics-service';
import { Statistics } from '../../../models/statistics/statistics';
import { UrlUtils } from '../../../utils/url.utils';

@Injectable()
export abstract class StatisticsService<TRequest, TData>
  implements IStatisticsService<TRequest, TData> {

  protected abstract getBaseUrl(): string;

  public getStatistics(request?: TRequest): Task<Statistics<TRequest, TData>> {
    return HttpTask.create(
      `${this.getBaseUrl()}?${UrlUtils.ToQueryString(request)}`)
      .get();
  }
}

